### PR TITLE
Reference Counting fixes and consolidation

### PR DESCRIPTION
**Pull request by charlesgregory**
Based on comments in #107 and on the https://github.com/blachlylab/rc_issues repo, its clear my reference counting implementation for `dhtslib` was incorrect. This pull request proposes to fix those issues but also centralize code for management (reference counting) of htslib pointers. 

### Main feature:
`HtslibMemory` is a templated struct that can be used to store any type of htslib pointer, reference count it, and free it using the appropriate function. Most of the implementation was inspired from `RefCounted` from D's `std.typecons`. 

~~Why not use `RefCounted`?~~

~~This allows us to easily use our "destroy" function of choice, and not have to explicitly specify that we want a `RefCounted!SAMRecord`.~~

~~Edit: We now use `RefCounted`.~~
Edit again: `RefCounted` cannot be used in `@safe` code so we now have our own reference counting implementation.

`SAMRecord` can use a `Bam1tPtr` which is an alias to `HtslibMemory!(bam1_t, bam_destroy)`. `SAMRecord` doesn't worry about any reference counting because as long as the `Bam1tPtr` is in scope memory accesses should be valid.

I have only first implemented `Bam1tPtr`, though I would intend to use this on most htslib structs and convert most dhtslib uses of htslib pointers to some templated form of `HtslibMemory`.Though I want comments and to nail down naming if we are to move forward with this. @jblachly let me know your thoughts. If you agree with this approach I will move forward aliasing other htslib pointer types.

### CODE REVIEW

File: source/dhtslib/memory.d [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        this.ptr = ptr;
+        this.refct = cast(size_t*)malloc(size_t.sizeof);
+        *this.refct = 1;
+    }
+
+    /// struct ctor
+    this(T data)
+    {
+        *this.ptr = data;
```
Comment (thewilsonator) [inline source/dhtslib/memory.d:31]: this will crash as `this.ptr is null`

File: source/htslib/package.d [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module htslib;
```
Comment (thewilsonator) [inline source/htslib/package.d:1]: this is a good idea, do it in a different pull request because it is independent of this change..

**Review by thewilsonator — COMMENTED** [review]
> Why not use RefCounted?

I would highly recommend doing so. Getting your own destroy method is fairly simple as creating a struct that wraps a e.g. `bam1_t` and in its destructor calls `bam_destroy`

File: source/dhtslib/memory.d (hunk 1 above)
Comment (charlesgregory) [reply source/dhtslib/memory.d:31]: You are right. I need more unittests. 
> > I would highly recommend doing so. Getting your own destroy method is fairly simple as creating a struct that wraps a e.g. bam1_t and in its destructor calls bam_destroy

I mainly chose against when making this as I couldn't see how to best use it and maintain flexibility. Though now I see I could turn `HtslibMemory` into a template the achieves the same result and uses `RefCounted`.

Comment (charlesgregory) [reply source/dhtslib/memory.d]: I have moved to using `RefCounted` for reference counting in 5e1a3ea. `HtslibMemory` is still a struct instead of a template as I previously considered above. Took me a minute of finagling to figure out how to still use the `HtslibMemory` type as the underlying `T *` type so I actually used the `RefCounted` implementation as a guide. I didn't realize that `alias this` could use a function.

File: source/dhtslib/memory.d [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            }
+        }
+    }
+
+    /// reference counted HtslibPtr
+    RefCounted!(HtslibPtr, RefCountedAutoInitialize.yes) rcPtr;
+
+    /// get underlying data pointer
+    @property nothrow @safe pure @nogc
+    ref inout(T*) truePtr() inout return
```
Comment (thewilsonator) [inline source/dhtslib/memory.d:46]: I think you also want to make the `scope` because while `HtslibPtr` is not copyable,  `T*` trivially is, and you done want a reference gained from `truePtr` to outlive `this`. This might already be handled by the `return` attribute though, I'm not an expert in D's safety features.

Comment (charlesgregory) [reply source/dhtslib/memory.d:47]: Looking into this, I am not completely sure how to achieve this either. I'm not sure either `scope` or `@live` are fleshed out enough to recognize when to keep `this` in scope if referencing a member's data. It would be nice if D had a `scope!this` I could attach to the returned reference. Either keeping `this` in scope or simply not compiling whenever I try to escape a reference to `this` with a longer lifetime than `this`.

Comment (thewilsonator) [reply source/dhtslib/memory.d:47]: when in doubt, ask on the learn forum

File: source/dhtslib/memory.d [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import htslib.hts_log;
+
+/// Template struct that performs reference
+/// counting on htslib pointers and destroys with specified function
+struct HtslibMemory(T, alias destroy)
+if(!isPointer!T && isSomeFunction!destroy)
+{
+    @safe:
+    /// Pointer Wrapper
+    struct HtslibPtr
```
Comment (thewilsonator) [inline source/dhtslib/memory.d:18]: you probably want to make this a `static struct` because it doesn't depend on anything from `HtslibMemory` (template parameters don't count)

Comment (charlesgregory) [reply source/dhtslib/memory.d]: done d26b393.

**Comment by charlesgregory** [comment]
Looks like a compiler flag may actually do what we want?
https://forum.dlang.org/post/prlpbrydqyvmotmsxxpl@forum.dlang.org

**Comment by jblachly** [comment]
> Looks like a compiler flag may actually do what we want? https://forum.dlang.org/post/prlpbrydqyvmotmsxxpl@forum.dlang.org

DIP1000 was inactive so long it is marked as superceded, and thus I wouldn't rely on any feature that requires `-dip1000` switch. AFAIK that switch enables escape analysis.

TBH, the language documentation is pretty poor or non existant when it comes to recent changes. For example, just finding documentation on what the function attribute `return` does is an exercise in insanity. https://dlang.org/spec/attribute.html#return links to https://dlang.org/spec/function.html#return-ref-parameters but none of that clarifies what the `return` function attribute (note: not parameter attribute) does, especially in the context of a member function, even IF escape analysis is enabled.
 
There is probably no good language solution at this time. If you can live with this, just go ahead and merge.

**Comment by jblachly** [comment]
Actually, I take that back. https://dlang.org/spec/function.html#return-ref-parameters section 19.16.6 item 2 reads:

"Struct non-static methods marked with the return attribute ensure the returned reference will not outlive the struct instance."

Of course, it's still up to us to assume and/or test that this section which is about `ref` functions still applies to functions that return pointers directly/explicitly.

It also doesn't indicate that `-dip1000` is required for this lifetime tracking, only for `scope` analysis

**Comment by charlesgregory** [comment]
I have some new code I will push soon that addresses all of this. I had to do a lot of stuff with attributes and make a custom ref counter to achieve this since `RefCounted` is not `@safe`. To enable a lot of the scope checking I have to mark functions with the `@live @safe` and use the `-dip1000` flag. My intention is to use `-dip1000` only for running the unittests and not as a default build setting though it has been indicated that eventually `-dip1000` will not be optional.

**Comment by jblachly** [comment]
At this time we don't want to force `-dip1000` on library consumers because i don't know how compat it is with the corpus of downstream code

**Comment by thewilsonator** [comment]
> At this time we don't want to force -dip1000 on library consumers because i don't know how compat it is with the corpus of downstream code

Note that that does not preclude ensuring that _this_ library builds with `-dip1000` successfully. 
Note also that this library building successfully with `-dip100` is a prerequisite for building any library or application dependant on this with `-dip1000`.

What I would recommend is: in a different PR try enabling `-dip1000` for CI, see what breaks (there's bound to be something) , and assuming that its not too difficult to fix (it might be tedious what with dealing with C libraries, fix them, merge that PR and rebase this one onto master and see if it still passes. If there are too many error to fix easily or you need to annotate the entirety of htslib and that auditing takes too much effort, don't worry too much about it, merge this PR and then fix them at leisure.

**Comment by charlesgregory** [comment]
So turns out don't need `@live` flag after all, my unit tests that are designed to trip the compiler scope rules still pass. Though must still use the `-dip1000` flag. I have added a unittest config for `-dip1000` to run these tests and added it to our github actions. 

It appears to work and be safe. The below function is prevented from compiling with `-dip1000`.
```D
auto testfun()
{
    auto rc1 = Bam1_t(bam_init1);
    assert(rc1.core.pos == 0);
    return rc1.getRef; /// return the pointer
}

Error: scope variable `rc1` may not be returned
```
I did have to do my own reference counting since `RefCounted` is not `@safe`. It is not thread-safe though that could be an easy-ish fix. Turn the `refct` into `shared` (and maybe the data pointer too?).

File: source/dhtslib/vcf/reader.d [hunk 5]
```diff
@@ -175,7 +175,8 @@ struct VCFReaderImpl(CoordSystem cs, bool isTabixFile)
         // * unpacking and
         // * destroying
         // its copy
-        return VCFRecord(this.vcfhdr, bcf_dup(this.b), this.MAX_UNPACK);
+        auto copiedBcf_1t = bcf_dup(this.b);
+        return VCFRecord(this.vcfhdr, copiedBcf_1t, this.MAX_UNPACK);
     }
```
Comment (charlesgregory) [inline source/dhtslib/vcf/reader.d:149]: I think this is the only line I had to change to get `-dip1000` to work for the rest of the package. But most of dhtslib code is not annotated as `@safe`. Not sure if that affects the compiler's ability to do checking for `scope`.

Comment (jblachly) [reply source/dhtslib/vcf/reader.d:149]: Make sure to add a separate CI compilation that uses `-dip1000` as Nic suggests

Comment (jblachly) [reply source/dhtslib/vcf/reader.d:149]: Nvm, I see you did

**Comment by charlesgregory** [comment]
Do we want to change the names of any of the `alias`s? Current format is just capitalization `bam1_t -> Bam1_t`, though I have also considered `bam1_t -> Bam1tPtr` and `bam1_t -> Bam1t`.

**Comment by thewilsonator** [comment]
I would recommend `Bam1` (or `Bam1Ptr` if `bam_t` is not a pointer).

**Comment by charlesgregory** [comment]
Aliases have been renamed to follow the following convention: drop the `_t` and convert to CamelCase. 
i.e. 
```
bcf_hdr_t -> BcfHdr
bam1_t -> Bam1
```

Since this has been addressed and tests are passing, we will go ahead and merge. Thanks @thewilsonator for your help!

**Comment by thewilsonator** [comment]
My pleasure.
